### PR TITLE
Hotkey config

### LINKED ISSUE #1311: Keyboard shortcut conflicting with Chrome "View Source" under macOS
> "Use a keyboard shortcut (Ctrl+Alt+U or ⌘⌥U) to open a shell in your primary instance" 

introduced in 1.0.0. This is basically a very nice feature, but it should not be hardcoded but configurable.

Web developers tend to use "View Source" in Chrome, which has the exact same shortcut and is now conflicting. Multipass has now precedence and the shortcut in Chrome is no longer working.

### PR DESCRIPTION
**Pull request by ricab**
Fixes #1311. Also introduces `get --raw`.

### CODE REVIEW

**Comment by ricab** [comment]
@Saviq what do you think about this? I am going a bit further than we discussed by tentatively placing things at the platform level, so that we can handle the mac case properly, rather than tell users they need to use ctrl for cmd and meta for ctrl. 

Refactoring needed to use current linux approach on win.

**Comment by Saviq** [comment]
How do you think they'll be inputting Command and Option? Cmd and Opt?

**Comment by ricab** [comment]
I was thinking something like that. Not sure about Opt. Here's what they use: https://support.apple.com/en-us/HT201236

**Comment by ricab** [comment]
But I suppose we can accept both opt and alt.

**Comment by ricab** [comment]
I have found that both QKeySequence and QHotKey have [a few rough corners](https://github.com/canonical/multipass/blob/a0cd5ff8bae5fe0020ceaa3bb485392d087b405f/src/platform/platform_shared.cpp#L33-L44). The altgr and meta issues are particularly unfortunate, because they don't work but there is no error whatsoever. The numpad issue too. 

The more I think about this, the more it feels like we should restrict the values to a subset that we know we can support well. Perhaps something like:

- one of [F1-F12], or
- one or more of [ctrl, alt, shift, meta], but not meta alone, plus any non-modifier key, excluding numpad, media keys, Fn, altgr, backspace, enter, caps lock, ... oh dear, how do we even document? We can't say [a-z0-9F1-F12] because that still excludes a bunch of symbols that work well but depend on keyboard layout (e.g. '.').

**Comment by ricab** [comment]
OK this should be about done now. Only need a final test on windows and probably some rebasing.

File: src/platform/platform_shared.cpp [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+// accepted with no warning)
+QString mpp::interpret_general_hotkey(const QString& val)
+{
+    auto sequence = QKeySequence{val};
+    auto ret = sequence.toString(QKeySequence::NativeText);
+
+    if (ret.isEmpty() && !sequence.isEmpty())
+        throw InvalidSettingsException(mp::hotkey_key, val, "Invalid key sequence");
+    if (sequence.count() > 1)
+        throw InvalidSettingsException(mp::hotkey_key, val, "Multiple key sequences are not supported");
```
Comment (Saviq) [inline src/platform/platform_shared.cpp:48]: Should we not try and compare `val` to `ret`? Or are they not guaranteed to be the same?

File: src/platform/platform_shared.h [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+#ifndef MULTIPASS_PLATFORM_SHARED_H
+#define MULTIPASS_PLATFORM_SHARED_H
+
+#include <QString>
+
+namespace multipass::platform
+{
+QString interpret_general_hotkey(const QString& val);
```
Comment (Saviq) [inline src/platform/platform_shared.h:25]: Why `_general_`?

File: src/utils/settings.cpp [hunk 3]
```diff
@@ -42,11 +44,17 @@ const auto client_root = QStringLiteral("client");
 const auto petenv_name = QStringLiteral("primary");
 const auto autostart_default = QStringLiteral("true");
 
+QString default_hotkey()
+{
+    return QKeySequence{mp::hotkey_default}.toString(QKeySequence::NativeText); // outcome depends on platform
```
Comment (Saviq) [inline src/utils/settings.cpp:49]: Oh so will this translate to "Cmd" and "Opt" on macOS?

**Comment by Saviq** [comment]
I wonder if we should display the hotkey in the GUI, then?

File: src/platform/platform_shared.cpp [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+// underscore, "_" is still interpreted the same as "-". IOW, "shift+-" == "shift+_" != "_" (just like "u" is the same
+// as "U").
+// 2) QHotKey fails to register some of the shortcuts we accept here (e.g. "Media Play").
+// 3) QKeySequence seems to have problems with AtlGr. QKeySequence("AltGr").toString() prints rubbish (that it does not
+// interpret back to mean the same thing). Unfortunately it is not enough to specify "ú" when that's what the layout
+// produces for AltGr+U. The Sequence "ú" is accepted and QHotKey registers it, but is gets triggered on "U" and not
+// "AltGr+U".
+// 4) There does not seem to be a way to specify numpad keys in QKeySequence (with or without numlock).
+// 5) meta only seems to work with other modifiers (e.g. ctrl+meta+x works, but meta+x doesn't even though it is
+// accepted with no warning)
```
[the remark this answers is not in the document]
Comment (ricab) [reply src/platform/platform_shared.cpp:39]: I am not sure TBH, those were conclusions from testing on Linux. I got some of those same results on the others, but I did not check each one on every platform. The items referring to `Meta` and `AltGr` have doubtful applicability on macOS (at least with the default keyboard).

File: src/platform/platform_shared.cpp (hunk 1 above)
Comment (ricab) [reply src/platform/platform_shared.cpp:48]: They're not guaranteed to be the same, "alt+ctrl+u" is a valid input that gets interpreted as "Ctrl+Alt+U".

File: src/platform/platform_shared.h (hunk 2 above)
Comment (ricab) [reply src/platform/platform_shared.h:25]: LOL, yeah indeed :smile: I'll remove that. 

When I started this was not in a separate header and was called by the actual platform functions to interpret the hotkey, so it was a way to distinguish.

File: src/utils/settings.cpp (hunk 3 above)
Comment (ricab) [reply src/utils/settings.cpp:49]: Yup :wink: Well, actually to ⌘⌥U

**Comment by ricab** [comment]
> I wonder if we should display the hotkey in the GUI, then?

Makes sense indeed.

File: src/platform/platform_shared.h (hunk 2 above)
Comment (ricab) [reply src/platform/platform_shared.h]: Done.

**Comment by Saviq** [comment]
Alright, nice!

bors merge

**Comment by Saviq** [comment]
Conflict on macOS:

```
CONFLICT (content): Merge conflict in src/platform/CMakeLists.txt
```

**Comment by ricab** [comment]
> Conflict on macOS:
> 
> ```
> CONFLICT (content): Merge conflict in src/platform/CMakeLists.txt
> ```

Hi, that's because bors does not grab the matching branch when staging, right?

**Pull request merged**